### PR TITLE
Updating CI with macOS build

### PR DESCRIPTION
**Pull request by Smit1603**
- Added steps to build LLVM/clang on macOS and works perfectly
- Errors are being faced while building and testing CppInterop hence needs to be fixed yet

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -106,6 +140,16 @@ jobs:
           distro \
           pytest
 
+    - name: Install deps on MacOS
+      if: runner.os == 'macOS'
+      run: |
+        # Install deps
+        ##sudo apt update
```
Comment (vgvassilev) [inline .github/workflows/ci.yml:147]: That seems irrelevant comment.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -106,6 +140,16 @@ jobs:
...
+        brew install git gcc gnupg python@3 
+        conda install -y -q -c conda-forge \
```
Comment (vgvassilev) [inline .github/workflows/ci.yml:149]: Can we get away from conda, and use only brew?

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -164,6 +209,63 @@ jobs:
           cmake --build . --target clang clang-repl --parallel $(nproc --all)
         fi
         cd ../../
+
+    - name: Build LLVM/Cling on macOS if the cache is invalid
```
Comment (vgvassilev) [inline .github/workflows/ci.yml:213]: This step is not really different from the step on unix. Can we merge them?

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -234,6 +339,58 @@ jobs:
         echo "LLVM_BUILD_DIR=$LLVM_BUILD_DIR" >> $GITHUB_ENV
         echo "CPLUS_INCLUDE_PATH=$CPLUS_INCLUDE_PATH" >> $GITHUB_ENV
 
+    - name: Build and Test/Install CppInterOp on macOS
```
Comment (vgvassilev) [inline .github/workflows/ci.yml:342]: Likewise.

**Review by vgvassilev — COMMENTED** [review]
This pull request looks promising. It uncovered issues in clang-repl and the build status of the OSX builds should be marked as expected to fail. That is, we should probably not run the tests, or run only the ones we know work.

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -56,14 +65,15 @@ jobs:
...
+
     - name: Save PR Info
       run: |
         mkdir -p ./pr
         echo ${{ github.event.number }} > ./pr/NR
         echo ${{ github.repository }} > ./pr/REPO
-
-        cling_on="${{ matrix.cling }}"
-        if [[ "${cling_on^^}" == "ON" ]]; then
+
```
[the remark this answers is not in the document]
Comment (Smit1603) [reply .github/workflows/ci.yml]: Actually made this changes as while running on macOS , I was facing bad substitution error and found that earlier command to make string uppercase wasn't available for macOS. Hence used this new command which is valid and running for both Linux and macOS

File: .github/workflows/ci.yml (hunk 1 above)
Comment (Smit1603) [reply .github/workflows/ci.yml]: Taken care of it

File: .github/workflows/ci.yml (hunk 2 above)
Comment (Smit1603) [reply .github/workflows/ci.yml]: Since its python libraries, so used pip command rather than brew.

File: .github/workflows/ci.yml (hunk 3 above)
Comment (Smit1603) [reply .github/workflows/ci.yml:209]: Yaa so I tried combining them and faced an unforeseen error where building of llvm got terminated unexpectedly. Hence looking for it.

File: .github/workflows/ci.yml (hunk 4 above)
Comment (Smit1603) [reply .github/workflows/ci.yml]: Yeah merged them, just a thing to be mentioned is lld linker is not supported on macOS. Hence i removed that flag and now default ld64 linker is being used in every environment. Is this change ok for our project??

Comment (vgvassilev) [reply .github/workflows/ci.yml]: This can be solved by adding a matrix switch to handle the platform-specific cmake options: https://github.com/vgvassilev/clad/blob/59fd672f54d171577857fe75ed9223b91d7902d1/.github/workflows/ci.yml#L93

File: unittests/CppInterOp/FunctionReflectionTest.cpp [hunk 6]
```diff
@@ -556,6 +558,7 @@ TEST(FunctionReflectionTest, IsVirtualMethod) {
   EXPECT_FALSE(Cpp::IsVirtualMethod(SubDecls[3])); // y()
 }
 
+#if !(__clang__ && __APPLE__)
 TEST(FunctionReflectionTest, JitCallAdvanced) {
```
Comment (vgvassilev) [inline unittests/CppInterOp/FunctionReflectionTest.cpp:562]:
```suggestion
#ifdef __APPLE__
TEST(FunctionReflectionTest, DISABLED_JitCallAdvanced) {
#else
TEST(FunctionReflectionTest, JitCallAdvanced) {
#endif
```

We should use this pattern everywhere.

Comment (Smit1603) [reply unittests/CppInterOp/FunctionReflectionTest.cpp]: Yes made changes according to this pattern everywhere

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -129,41 +175,59 @@ jobs:
...
                 -DCLANG_ENABLE_ARCMT=OFF                      \
                 -DCLANG_ENABLE_FORMAT=OFF                     \
                 -DCLANG_ENABLE_BOOTSTRAP=OFF                  \
                 ../llvm
-          cmake --build . --target clang clang-repl --parallel $(nproc --all)
+          if [[ "${os}" == "macos*" ]]; then
+            cmake --build . --target clang clang-repl --parallel $(sysctl -n hw.physicalcpu)
+          else
+            cmake --build . --target clang clang-repl --parallel $(nproc --all)
+          fi
```
Comment (vgvassilev) [inline .github/workflows/ci.yml:227]: Can you make this somewhere in the settings of the build and then just use ncpus instead of if-stmts everywhere?

```
if [[ "${os}" == "macos*" ]]; then
   ncpus=$(sysctl -n hw.physicalcpu)
else
   ncpus=$(nproc --all)
fi
```

Comment (alexander-penev) [reply .github/workflows/ci.yml:227]: Maybe we should use hw.ncpu instead hw.physicalcpu

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -1,9 +1,9 @@
 name: Main
 on:
-  pull_request:
-    branches: [main]
+  # pull_request:
+  #   branches: [main]
```
Comment (vgvassilev) [inline .github/workflows/ci.yml:4]: Why did you comment this section?

File: .github/workflows/ci.yml (hunk 7 above)
Comment (Smit1603) [reply .github/workflows/ci.yml]: - Made an environment variable to store number of cpu cores and replaced hw.physicalcpu to hw.ncpu

**Review by vgvassilev — APPROVED** [review]
LGTM!
